site/search/searchHooks.ts

```diff
@@ -297,3 +243,203 @@ export const useTopicTagGraph = () => {
 
     return tagGraph
 }
+
+/**
+ * Hook that manages search state via URL search params.
+ *
+ * Key design principle: URL is the source of truth. State is derived
+ * synchronously from URL params.
+ */
+export function useSearchParamsState(): {
+    state: SearchState
+    actions: SearchActions
+} {
+    const [searchParams, setSearchParams] = useSearchParams()
+
+    // Derive state from URL
+    const state = useMemo(
+        () => searchParamsToState(searchParams),
+        [searchParams]
+    )
+
+    // Helper to update params atomically
+    const updateParams = useCallback(
+        (updater: (current: SearchState) => SearchState) => {
+            setSearchParams((prev) => {
+                const currentState = searchParamsToState(prev)
+                const newState = updater(currentState)
+                return stateToSearchParams(newState)
+            })
+        },
+        [setSearchParams]
+    )
+
+    const actions = useMemo<SearchActions>(
+        () => ({
+            setQuery: (query: string) => {
+                updateParams((s) => ({ ...s, query: query.trim() }))
+            },
+
+            addCountry: (country: string) => {
+                updateParams((s) => {
+                    const exists = s.filters.some(
+                        (f) =>
+                            f.type === FilterType.COUNTRY && f.name === country
```

Comment:
Is there a reason why we constrain ourselves to this generic `filters` property with `filter.type` instead of having multiple properties like `countryFilters`, `topicFilters` etc. on the `SearchState`? We are not creating filter types dynamically, so we can be more concrete and make the state manipulation easier. And then it might be possible to use sets for the filter values instead of arrays and thus avoid the `exists` checks etc.